### PR TITLE
[Model] Support multiple images for qwen-vl

### PR DESCRIPTION
**Pull request by alex-jw-brooks**
This PR finishes exposing multi-image support for Qwen-VL (not Qwen2) as follow-up to https://github.com/vllm-project/vllm/pull/8029. 

### Multi-image offline inference example (.generate)
```python
from vllm import LLM, SamplingParams
from vllm.multimodal.utils import fetch_image

question = "Can you compare these images?"
image_urls = [
    "https://upload.wikimedia.org/wikipedia/commons/d/da/2015_Kaczka_krzy%C5%BCowka_w_wodzie_%28samiec%29.jpg",
    "https://upload.wikimedia.org/wikipedia/commons/7/77/002_The_lion_king_Snyggve_in_the_Serengeti_National_Park_Photo_by_Giles_Laurent.jpg",
]

llm = LLM(
    model="Qwen/Qwen-VL-Chat",
    trust_remote_code=True,
    limit_mm_per_prompt={"image": len(image_urls)},
    dtype="half"
)


get_img_prompt = lambda img_num: f"Picture {img_num}: <img></img>\n"
placeholders = "".join(get_img_prompt(i) for i, _ in enumerate(image_urls, start=1))
prompt = f"<|im_start|>user\n{placeholders}{question}\n<|im_end|>\n<|im_start|>assistant\n"

sampling_params = SamplingParams(temperature=0.2, max_tokens=64)

outputs = llm.generate({
        "prompt": prompt,
        "multi_modal_data": {
            "image": [fetch_image(url) for url in image_urls]
        },
    },
    sampling_params=sampling_params
)

for o in outputs:
    generated_text = o.outputs[0].text
    print(generated_text)
```
```
The two images shown are different. The first image is of a male mallard duck swimming in a body of water, while the second image is of a male lion resting in the tall grass. The two images are unrelated and cannot be compared.<|im_end|>
```

### Chat example:
Image numbering is already handled properly in the chat utils, so no extra changes needed there.

Server:
```
python -m vllm.entrypoints.openai.api_server \
    --device cuda \
    --model Qwen/Qwen-VL-Chat \
    --api-key token-abc123 \
    --chat-template examples/template_chatml.jinja \
    --tokenizer Qwen/Qwen-VL-Chat \
    --dtype=half \
    --limit-mm-per-prompt image=2 \
    --trust-remote-code &
```
Client:
```python
from openai import OpenAI

client = OpenAI(base_url="http://localhost:8000/v1", api_key="token-abc123")

completion = client.chat.completions.create(
  model="Qwen/Qwen-VL-Chat",
  messages=[
    {
        "role": "user", "content": [
          {"type": "image_url", "image_url": {"url": "https://upload.wikimedia.org/wikipedia/commons/d/da/2015_Kaczka_krzy%C5%BCowka_w_wodzie_%28samiec%29.jpg"}},
          {"type": "image_url", "image_url": {"url": "https://upload.wikimedia.org/wikipedia/commons/7/77/002_The_lion_king_Snyggve_in_the_Serengeti_National_Park_Photo_by_Giles_Laurent.jpg"}},
          {"type": "text", "text": "Can you compare these images?"},
        ]
    }
  ]
)

print(completion.choices[0].message)
```
```
ChatCompletionMessage(content='The images shown are different and cannot be compared. Image 1 is a duck swimming in water, and Image 2 is a lion lying in the grass. They are completely different in both content and subject.<|im_end|>\n', refusal=None, role='assistant', function_call=None, tool_calls=[])
```

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by alex-jw-brooks** [comment]
FYI @DarkLight1337 since you reviewed the main PR for qwen-vl! 🙂

**Comment by DarkLight1337** [comment]
Could you add a test to verify the behavior of this model?

After #8201 is merged, can you update the example script for multi-image input to include this model as well?

**Comment by alex-jw-brooks** [comment]
Yup, definitely! I had recently started working on some stuff around the image processor to allow kwargs to be passed through as overrides for stuff pulled from the HF config - but before that, I was planning to open a PR to add some common utils for building an `InputContext` and some dummy data for different multimodal models to test correctness of custom input mappers / processors. I think it would be nice to have fast tests that only test the preprocessing correctness that can run without loading the full models.

I had started writing those tests for this model already - do you have any thoughts on adding those here to test this PR? Happy to do that and then submit a follow-up refactoring common stuff out + adding stuff for other models, or if you think it would be better, just adding an end-to-end test here for now and adding preprocessing tests in a separate PR

**Comment by DarkLight1337** [comment]
> I had started writing those tests for this model already - do you have any thoughts on adding those here to test this PR? 

Sure!

> Happy to do that and then submit a follow-up refactoring common stuff out + adding some tests for other models, or if you think it would be better, just adding an end-to-end test here for now and adding preprocessing tests in a separate PR

I'm currently working on #7820, perhaps it would be best to refactor the tests after that.

**Comment by alex-jw-brooks** [comment]
Perfect, I'll add them to the test file to this model for now, and wait until after that's merged before making stuff common and adding similar tests for some other models 😄

File: examples/offline_inference_vision_language_multi_image.py
Comment (DarkLight1337) [inline examples/offline_inference_vision_language_multi_image.py:105]:
```suggestion
def run_generate(model: str, question: str, image_urls: List[str]):
```

File: examples/offline_inference_vision_language_multi_image.py
Comment (alex-jw-brooks) [reply examples/offline_inference_vision_language_multi_image.py]: Absolutely 😆 just did!

**Comment by DarkLight1337** [comment]
Can we have an end-to-end correctness test for multi-image input just like the other models?

**Review by DarkLight1337 — APPROVED** [review]
LGTM, thanks for your help again!

**Comment by DarkLight1337** [comment]
Can you merge `main` into this branch to fix the CI failure?

**Comment by alex-jw-brooks** [comment]
No problem, thanks a lot for the quick reviews! Just did 🙂

**Comment by DarkLight1337** [comment]
Sorry I forgot about this, Qwen2-VL has just been released so you have to update the examples file.

**Comment by alex-jw-brooks** [comment]
No worries at all, that's great news! Just rebased 😄

**Review by DarkLight1337 — APPROVED** [review]

**Pull request merged**
